### PR TITLE
Add new condition DNC to campaigns

### PR DESCRIPTION
**Pull request by tomekkowalczyk**
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" for all features, enhancements and bug fixes (until 3.3.0 is released) 
| Bug fix?                               | no
| New feature?                           | yes
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      |  
| Related developer documentation PR URL | 
| Issue(s) addressed                     |  

#### Description:
The functionality allows you to handle contacts from the level of the campaign condition that do not have or have the DNC flag and appropriate DNC channels, eg Emails.
image

image

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/9723"></a>

### CODE REVIEW

File: app/bundles/LeadBundle/EventListener/CampaignSubscriber.php [hunk 1]
```diff
@@ -47,49 +49,31 @@ class CampaignSubscriber implements EventSubscriberInterface
 {
     const ACTION_LEAD_CHANGE_OWNER = 'lead.changeowner';
 
-    /**
```
Comment (kuzmany) [inline app/bundles/LeadBundle/EventListener/CampaignSubscriber.php]: These comments should not be removed

File: app/bundles/LeadBundle/Translations/en_US/messages.ini [hunk 2]
```diff
@@ -261,6 +261,8 @@ mautic.lead.lead.events.campaigns.expression="Expression"
 mautic.lead.lead.events.campaigns.date="Date"
 mautic.lead.lead.events.owner="Contact owner"
 mautic.lead.lead.events.owner_descr="Condition based on a contact owner."
+mautic.lead.lead.events.condition_donotcontact="Has flag DNC"
```
Comment (kuzmany) [inline app/bundles/LeadBundle/Translations/en_US/messages.ini]: Just thinking about better naming.
Has Do Not Contact flag
Maybe @RCheesley as native speaker could take look on it.

File: app/bundles/LeadBundle/Form/Type/CampaignEventLeadDNCType.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
+<?php
+
+/*
+ * @copyright   2019 Mautic Contributors. All rights reserved
```
Comment (kuzmany) [inline app/bundles/LeadBundle/Form/Type/CampaignEventLeadDNCType.php]: This should be 2021

**Review by kuzmany — CHANGES_REQUESTED** [review]
I like this PR
I've address few minor comments.
But I think this switch Has DNC flag is not necessary.

image

Campaign condition send contact to true/false patch based on If contact is contactable.

Then you are able to cover both situation - has DNC or has not DNC

File: app/bundles/LeadBundle/Translations/en_US/messages.ini (hunk 2 above)
Comment (RCheesley) [reply app/bundles/LeadBundle/Translations/en_US/messages.ini]: I would suggest 'Marked as DNC' if we are trying to show why a contact is not being contacted, their DNC status or to confirm the successful application of the DNC status.

Comment (kuzmany) [reply app/bundles/LeadBundle/Translations/en_US/messages.ini]: that sounds perfect 👍

**Comment by tomekkowalczyk** [comment]
@kuzmany Could you explain how you want to handle contacts with or without the DNC flag by getting rid of the "Has flag DNC" button?

**Comment by RCheesley** [comment]
@tomekkowalczyk following up on the feedback I think that what @kuzmany was suggesting is that if you have the condition of 'do they have the DNC flag' then you would use the green (yes) pathway in the affirmative (they have the flag set) and the red (no) pathway in the negative (they do not have the flag set).

Adding another switch in the setting for the condition would add complexity where it should not be needed, as we already have a yes/no from the campaign builder.

Also, a conflict to address please :)

**Comment by tomekkowalczyk** [comment]
@RCheesley The conflict was resolved

**Comment by tomekkowalczyk** [comment]
@RCheesley I created new PR https://github.com/mautic/mautic/pull/12099 with 5.x line

**Pull request closed without merging**